### PR TITLE
SystemParamBuilder - Support dynamic system parameters

### PR DESCRIPTION
**Pull request by chescock**
# Objective

Support building systems with parameters whose types can be determined at runtime.  

## Solution

Create a `DynSystemParam` type that can be built using a `SystemParamBuilder` of any type and then downcast to the appropriate type dynamically.

## Example

```rust
let system = (
    DynParamBuilder::new(LocalBuilder(3_usize)),
    DynParamBuilder::new::<Query<()>>(QueryParamBuilder::new(|builder| {
        builder.with::<A>();
    })),
    DynParamBuilder::new::<&Entities>(ParamBuilder),
)
    .build_state(&mut world)
    .build_system(
        |mut p0: DynSystemParam, mut p1: DynSystemParam, mut p2: DynSystemParam| {
            let local = p0.downcast_mut::<Local<usize>>().unwrap();
            let query_count = p1.downcast_mut::<Query<()>>().unwrap();
            let entities = p2.downcast_mut::<&Entities>().unwrap();
        },
    );
```

### CODE REVIEW

File: crates/bevy_ecs/src/system/builder.rs [hunk 1]
```diff
@@ -382,4 +411,33 @@ mod tests {
         let result = world.run_system_once(system);
         assert_eq!(result, 5);
     }
+
+    #[test]
+    fn dyn_builder() {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/builder.rs:416]: I would like another test verifying that the unhappy path panics.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
This needs an end-to-end example in the form of a doc-test, probably on `DynSystemParam` to be properly taught to users. Handy though, and I appreciate how thorough you've been with the design and safety invariants.

File: crates/bevy_ecs/src/system/builder.rs (hunk 1 above)
Comment (chescock) [reply crates/bevy_ecs/src/system/builder.rs:416]: I've got the `assert!(p0.downcast_mut::<Query<()>>().is_none());` there to verify that you can't downcast to the wrong type.  (It returns `Option`, so there doesn't need to be a panic.)  Do you want that moved to a separate test?

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/builder.rs:416]: Hmm, it can stay where it is I suppose.

**Review by alice-i-cecile — APPROVED** [review]
Nice docs!

File: crates/bevy_ecs/src/system/system_param.rs [hunk 2]
```diff
@@ -1638,6 +1639,253 @@ unsafe impl<T: ?Sized> SystemParam for PhantomData<T> {
...
+        // SAFETY:
+        // - The caller ensures the world has access for the underlying system param,
+        //   and since the downcast succeeded, the underlying system param is T.
+        // - The caller ensures the `world` matches.
+        unsafe { T::get_param(&mut state.0, system_meta, world, change_tick) }
+    })
+}
+
+/// The [`SystemParam::State`] for a [`DynSystemParam`].
+pub struct DynSystemParamState(Box<dyn DynParamState>);
```
Comment (cBournhonesque) [inline crates/bevy_ecs/src/system/system_param.rs:1789]: just a nit, but the names being so similar make this a bit hard to follow sometimes.

They basically represent the same thing, but one is the trait and the other the boxed wrapper that can be used in the DynParamBuilder closure?

File: crates/bevy_ecs/src/system/system_param.rs [hunk 3]
```diff
@@ -1638,6 +1639,253 @@ unsafe impl<T: ?Sized> SystemParam for PhantomData<T> {
...
+    /// This is used to apply [`Commands`] during [`apply_deferred`](crate::prelude::apply_deferred).
+    ///
+    /// [`Commands`]: crate::prelude::Commands
+    fn apply(&mut self, system_meta: &SystemMeta, world: &mut World);
+
+    /// Queues any deferred mutations to be applied at the next [`apply_deferred`](crate::prelude::apply_deferred).
+    fn queue(&mut self, system_meta: &SystemMeta, world: DeferredWorld);
+}
+
+/// A wrapper around a [`SystemParam::State`] that can be used a a trait object in a [`DynSystemParam`].
```
Comment (cBournhonesque) [inline crates/bevy_ecs/src/system/system_param.rs:1818]:
```suggestion
/// A wrapper around a [`SystemParam::State`] that can be used as a trait object in a [`DynSystemParam`].
```

**Review by cBournhonesque — APPROVED** [review]

File: crates/bevy_ecs/src/system/system_param.rs (hunk 2 above)
Comment (chescock) [reply crates/bevy_ecs/src/system/system_param.rs:1789]: Yeah, the concepts are similar enough that I had trouble coming up with good names.  If you can thing of ones that would be more clear, I'm happy to create or review a PR to rename them!

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1679 if you'd like to help out.
